backends/dpdk/dpdkHelpers.cpp

```diff
@@ -346,6 +437,166 @@ bool ConvertStatementToDpdk::preorder(const IR::AssignmentStatement *a) {
+                    for (auto it : typeheader->fields) {
+                        IR::ID name(params->member.name + "." + it->name);
+                        auto m = new IR::Member(new IR::PathExpression(IR::ID("h")), name);
+                        components.push_back(m);
+                    }
+                } else {
+                    components.push_back(params);
+                }
+            }
+        }
+    }
+}
+
+/* This function moves the hash parameter fields to metadata structure if they happen
+   to come from different header/metadata structures or are not consecutively laid in
+   the structure they are coming from
+*/
+void ConvertStatementToDpdk::updateMdStrAndGenInstr(const IR::Argument* field,
+                                               IR::Vector<IR::Expression>& components) {
+    if (auto s = field->expression->to<IR::StructExpression>()) {
+        for (auto field1 : s->components) {
+            if (auto params = field1->expression->to<IR::Member>()) {
+                auto typeInfo = typemap->getType(params);
+                if (typeInfo->is<IR::Type_Header>()) {
+                    auto typeheader = typeInfo->to<IR::Type_Header>();
+                    for (auto it : typeheader->fields) {
+                        IR::ID name(refmap->newName(params->member.name + "." + it->name));
+                        IR::ID fldName(refmap->newName(typeheader->name.toString() + "_" +
+                                                       it->name));
+                        auto rt = new IR::Member(new IR::PathExpression(IR::ID("h")), name);
+                        auto lt = new IR::Member(new IR::PathExpression(IR::ID("m")), fldName);
+                        BUG_CHECK(metadataStruct, "Metadata structure missing unexpectedly!");
+                        metadataStruct->fields.push_back(new IR::StructField(fldName, it->type));
+                        add_instr(new IR::DpdkMovStatement(lt, rt));
+                        components.push_back(lt);
+                    }
+                } else {
+                    IR::ID name(refmap->newName(getHdrMdStrName(params) + "_" +
+                                                params->member.name));
+                    BUG_CHECK(metadataStruct, "Metadata structure missing unexpectedly!");
+                    metadataStruct->fields.push_back(new IR::StructField(name, params->type));
+                    auto lt = new IR::Member(new IR::PathExpression(IR::ID("m")), name);
+                    add_instr(new IR::DpdkMovStatement(lt, params));
+                    components.push_back(lt);
+                }
+            }
+        }
+    }
+}
+
+/* This function returns the header/metadata structure name */
+cstring ConvertStatementToDpdk::getHdrMdStrName(const IR::Member* mem) {
+    cstring sName = "";
+    if ((mem != nullptr) && (mem->expr != nullptr) &&
+       (mem->expr->type != nullptr)) {
+        if (auto st = mem->expr->type->to<IR::Type_Header>()) {
```

Comment:
can this refer to a local variable with type header?

Reply:
No. The argument passed to this function is either a member or a metadata structure.